packages/flutter/lib/src/painting/image_cache.dart

```diff
@@ -185,3 +190,21 @@ class _CachedImage {
   final ImageStreamCompleter completer;
   final int sizeBytes;
 }
+
+class _PendingImage {
+  _PendingImage(this.completer);
+
+  final ImageStreamCompleter completer;
+  ImageListener _listener;
+
+  void addListener(ImageListener listener) {
+    assert(_listener == null);
+    _listener = listener;
+    completer.addListener(listener);
+  }
+
+  void removeListener() {
+    completer.removeListener(_listener);
+    _listener = null;
+  }
+}
```

Comment:
why do we use an add/remove pattern if we can only have one listener? should we just pass it into the constructor and make it final? Or use a getter/setter pair? or an "update" method that takes a listener or null?